File: Dynamic_Programming_Old/canSum-howSum-bestSum/howSum.cpp
#include <bits/stdc++.h>
using namespace std;
using ull = unsigned long long int;

map<int, ull> memo = {};

void pMemo()
{
    for (auto e : memo)
        cout << e.first << "\t" << e.second;
}
void pVector(vector<int>* v)
{
    cout << &v;
    for (int i = 0; i < v->size(); i++)
        cout << v->at(i);
}

void pVector(vector<int>& v)
{
    for (int i = 0; i < v.size(); i++)
        cout << v.at(i) << " ";

    cout << endl;
}

vector<int>* make(int m = 0, int n = 0)
{
    return new vector<int>(m, n);
}

// bool howSum(int targetSum, vector<int> numbers)
// {
//     static vector<int> seq;

//     if (targetSum == 0)
//     {
//         pVector(seq);
//         seq.clear();
//         return true;
//     }
//     if (targetSum < 0)
//         return false;

//     for (auto num : numbers)
//     {
//         seq.push_back(num);

//         int remain = targetSum - num;

//         if (howSum(remain, numbers) == true)
//             return true;
//         else
//             seq.pop_back();
//     }

//     return false;
// }

bool howSum(int targetSum, vector<int> numbers)
{
    static vector<int> seq;

    if (targetSum == 0)
    {
        pVector(seq);
        return true;
    }
    if (targetSum < 0)
        return false;

    for (auto num : numbers)
    {
        seq.push_back(num);

        int remain = targetSum - num;

        if (howSum(remain, numbers) == true)
            seq.pop_back();
        else
            seq.pop_back();
    }
}

int main()
{
    // system("cls");
    howSum(7, { 3, 1 });
    // howSum(7, { 7, 7 });
    // howSum(7, {5, 3});
    // howSum(7, {3, 2});
    // howSum(7, {7, 5, 4, 3});
    // howSum(7, {4, 2});
    // howSum(6, {3, 2, 1});
    // howSum(8, {3, 5, 2});
    // howSum(6, {7, 6, 3});
}

File: Dynamic_Programming_Old/howSum.js
// arr = [1,2,3];
// console.log(arr);
// arr = [...arr, 5];
// console.log(arr);

// arr.push(9);
// console.log(arr);




// console.clear();
// m = 2;
// n = 16;
// r = ()=> 2;
// arr = Array(m).fill(Array(n).fill(4));

// console.log(arr);


// for(i of arr){
//     console.log(JSON.stringify(i));
// }

const howSum = (targetSum, numbers, memo = {}) => {
    if(targetSum in memo) return memo[targetSum];
    if(targetSum === 0) return [];
    if(targetSum <0) return null;

    for(let num of numbers) {
        const remainder  = targetSum - num;
        const remainderResult = howSum(remainder, numbers, memo);

        if(remainderResult != null) {
            memo[targetSum] = [...remainderResult, num];
            return memo[targetSum];
        }
    }

    memo[targetSum] = null;
    return null;
}

console.clear();
console.log(howSum(7, [2, 3]));
console.log(howSum(7, [5,3,4,7]));
console.log(howSum(7, [2,4]));
console.log(howSum(8, [2,3,5]));
console.log(howSum(300, [7,14]));